org.eclipse.lsp4e/src/org/eclipse/lsp4e/MarkerAttributeComputer.java

```diff
@@ -0,0 +1,116 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2017 Red Hat Inc. and others.
+ * This program and the accompanying materials are made
+ * available under the terms of the Eclipse Public License 2.0
+ * which is available at https://www.eclipse.org/legal/epl-2.0/
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *
+ * Contributors:
+ *  Rubén Porras Campo (Avaloq) - extracted to separate file
+ *******************************************************************************/
+package org.eclipse.lsp4e;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+
+import org.eclipse.core.resources.IMarker;
+import org.eclipse.core.resources.IResource;
+import org.eclipse.jdt.annotation.NonNull;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.jface.text.BadLocationException;
+import org.eclipse.jface.text.IDocument;
+import org.eclipse.lsp4j.Diagnostic;
+import org.eclipse.lsp4j.Range;
+
+/**
+ * A class that computes the attributes of a
+ * {@link org.eclipse.core.resources.IMarker}.
+ *
+ * It can be extended by sub-classing it to add additional attributes to the
+ * ones already set by this class.
+ * <p>
+ * When doing so {@link #computeMarkerAttributes(IResource)} and
+ * {@link #computeMarkerAttributes(IDocument, Diagnostic)} should reuse the
+ * basis implementation and only add attributes from the returned map.
+ *
+ * Implementations can also customize {@link #attributeCount()} to optimize the
+ * size of the attribute map when it is created.
+ *
+ * <p>
+ * Implementation detail: For performance reasons we allow subclasses to add
+ * attributes directly to the map returned by
+ * {@link #computeMarkerAttributes(IDocument, Diagnostic)}. The attributes
+ * returned by {@link #computeMarkerAttributes(IResource)} as well as an
+ * attribute containing the language server ID are added as well to this map.
+ * <p>
+ * {@link #computeMarkerAttributes(IResource)} returns an immutable empty
+ * collection, which cannot be modified.
+ */
+public class MarkerAttributeComputer {
+
+	// Specific marker attributes defined by LSP4E
+	public static final String LSP_DIAGNOSTIC = "lspDiagnostic"; //$NON-NLS-1$
+
+	/**
+	 * Defines the expected number of attributes that will be computed, it is used
+	 * to optimize size of the map at creation time.
+	 *
+	 * @return the expected number of attributes that will be computed
+	 */
+	public int attributeCount() {
+		return 8;
+	}
+
+	/**
+	 * Computes the attributes of a marker for the given resource.
+	 *
+	 * @param resource,
+	 *            the {@link Resource} where this marker will be created
+	 * @return a map with the marker attributes
+	 */
+	public Map<String, Object> computeMarkerAttributes(@NonNull IResource resource) {

```

Comment:
It could look like one, but it is not a mistake, it is by intention.

We have marker attributes that depend only on the resource (an not the diagnostic itself), like for example the name of the resource, the extension, some elements of the path, etc (we also have elements which are not part of the path, and they are not that cheap to compute).
That is why we can calculate these values only once before the loop that iterates over the diagnostics and just append the information that we calculated. By doing it like this, we avoid recalculating values which are always equal for a resource.

Reply:
I'm not certain it belongs to LSP4E then if it doesn't depend on LS result. I think this would better be a resource listener that amends the markers when they're created (with a resource listener); or maybe it's better if you can plug this extra code in the existing method that takes document as input, by looking up the resource for the document and adding the necessary attributes if it makes sense.
But I don't think a dedicated method here is the right place.

Reply:
OK, I have take it out from the PR, I will take a look at how to use a resource listener. I do not think I will use it, but I looked and could not find it either, so I would like to ask if you know with which API can one look up the resource for the document, something like what is it done in org.eclipse.lsp4e.LSPEclipseUtils.getFile(IDocument)? That looks quite expensive for something one would like to do for each marker.

Reply:
If you feel more comfortable with it, you can also make the computeMarkerAttributes method take the resource as input as well, so it would be `Map<String, Object> computeMarkerAttributes(Diagnostic diagnostic, IDocument document, Optional<IResource> resource)`. What do you think?

Reply:
Hi Mickael,

the problem with `Map<String, Object> computeMarkerAttributes(Diagnostic diagnostic, IDocument document, Optional<IResource> resource)` is that I would need to recompute the resource markers for each marker.

I took some basic statistics, a normal project for us has 13k markers. For resources having markers the average is 4 markers/resource, and some outliers have more than 100 markers. Since I new that (before doing the numbers) I optimized it to be computed only once. Because of this I personally like more the single method which results are calculated once and then just propagated.

I also looked at the version with the resource listener. It will also perform worse, for the 3k resources, I will need to process the resource change (probably not a big deal), but then call `IMarker.setAttributes(Map<String, ? extends Object>)` on each marker. The implementation is using a workspace operation, so this means 13k extra workspace operations to set the "resource attributes".

I do not know which version performs better, probably the first, so yes, between these two I would chose the first, but I prefer the version with the extra method I had before, it is functionally equivalent but much faster ;)

I know 13k markers is a lot, but we do not have control over the code base, so we strive to be as efficient as possible.

I would propose that if you do not want the extra method for the resource calculation, we leave it as in this PS, I look at the listener and if it performs poorly we can always add the extra method I had, if you do not mind much, I would rather prefer having it, it is for me less effort that any other option and I already know it performs better.


Regards

Reply:
OK, I'd be fine merging this patch.
However, if you want to see resource involved, it would be possible in LSPDiagnosticsToMarkers to first group diagnostics by URI so you would resolve resource & document only once for each URI and pass them to the MarkerAttributeComputer without too much extra overhead.
I let you decide whether you prefer to implement that suggestion, or if you prefer we merge the patch as is.

Reply:
Hi Mickael,

sorry, I do not follow you when you propose to "first group diagnostics by URI"? 

I think LSPDiagnosticsToMarkers already process one resource at a time when updating markers, doesn't it? That is why `updateMarkers(PublishDiagnosticsParams, IResource)` has the resource as input already. In my previous PR I already improved the code to resolve the document of the resource only once for each URI. That is why calculating the attributes for resource it is very easy with low overhead (just one extra call).

When preparing my answer I have notice that the approach of a resource listener to add attributes would not play nicely with LPSP4E, because the updateMarker code 
```
	protected void updateMarker(@NonNull Map<String, Object> targetAttributes, @NonNull IMarker marker) {
		try {
			if (!targetAttributes.equals(marker.getAttributes())) {
				marker.setAttributes(targetAttributes);
			}
		} catch (CoreException e) {
			LanguageServerPlugin.logError(e);
		}
	}

```
just compares the attributes known by the MarkerAttributeComputer and if they are different than the existing ones, they are just reset. Yes, the listener could set them again, but that is a lot of overhead and unnecessary things going on. So I think somehow we need to have the resource as part of the computation of MarkerAttributeComputer.

What do you think if we merge this PR as it is, then I open a new one to discuss take a look at the different possibilities? I can then submit two or three different PR and we can easily compare them.

I believe we can quickly settle to one of the PRs once I push them for review, and if you do not plan a release right now, I think that is a nice way going forward.

Regards